airbyte-integrations/connectors/source-mixpanel/source_mixpanel/source.py

```diff
@@ -0,0 +1,838 @@
+    API Docs: https://developer.mixpanel.com/reference/cohorts
+    Endpoint: https://mixpanel.com/api/2.0/cohorts/list
+
+    [{
+        "count": 150
+        "is_visible": 1
+        "description": "This cohort is visible, has an id = 1000, and currently has 150 users."
+        "created": "2019-03-19 23:49:51"
+        "project_id": 1
+        "id": 1000
+        "name": "Cohort One"
+    },
+    {
+        "count": 25
+        "is_visible": 0
+        "description": "This cohort isn't visible, has an id = 2000, and currently has 25 users."
+        "created": "2019-04-02 23:22:01"
+        "project_id": 1
+        "id": 2000
+        "name": "Cohort Two"
+    }
+    ]
+
+    """
+
+    data_field = None
+    primary_key = "id"
+
+    def path(self, **kwargs) -> str:
+        return "cohorts/list"
+
+
+class FunnelsList(MixpanelStream):
+    """List all funnels
+    API Docs: https://developer.mixpanel.com/reference/funnels#funnels-list-saved
+    Endpoint: https://mixpanel.com/api/2.0/funnels/list
+    """
+
+    primary_key = "funnel_id"
+    data_field = None
+
+    def path(self, **kwargs) -> str:
+        return "funnels/list"
+
+
+class DateSlicesMixin:
+    def stream_slices(
+        self, sync_mode, cursor_field: List[str] = None, stream_state: Mapping[str, Any] = None
+    ) -> Iterable[Optional[Mapping[str, Any]]]:
+        date_slices = []
+
+        # use the latest date between self.start_date and stream_state
+        start_date = self.start_date
+        if stream_state:
+            # Remove time part from state because API accept 'from_date' param in date format only ('YYYY-MM-DD')
+            # It also means that sync returns duplicated entries for the date from the state (date range is inclusive)
+            stream_state_date = datetime.fromisoformat(stream_state["date"]).date()
+            start_date = max(start_date, stream_state_date)
+
+        # use the lowest date between start_date and self.end_date, otherwise API fails if start_date is in future
+        start_date = min(start_date, self.end_date)
+
+        # move start_date back <attribution_window> days to sync data since that time as well
+        start_date = start_date - timedelta(days=self.attribution_window)
+
+        while start_date <= self.end_date:
+            end_date = start_date + timedelta(days=self.date_window_size - 1)  # -1 is needed because dates are inclusive
+            date_slices.append(
+                {
+                    "start_date": str(start_date),
+                    "end_date": str(min(end_date, self.end_date)),
+                }
+            )
+            # add 1 additional day because date range is inclusive
+            start_date = end_date + timedelta(days=1)
+
+        # reset reqs_per_hour_limit if we expect less requests (1 req per stream) than it is allowed by API reqs_per_hour_limit
+        if len(date_slices) < self.reqs_per_hour_limit:
+            self.reqs_per_hour_limit = 3600  # 1 query per sec
```

Comment:
I still don't understand what this logic is doing. If we do have less than the limit, why would we want to suddenly send 1 QPS? Which is much more requests than the supposedly rate limit.

Reply:
Is this because Mixpanel bucket request by the hour? e.g. they don't care if we send 399 request in the first minute and the last request in the last minute as long as we are less than 400 and we want to send the request as fast as possible?

Reply:
yes, exactly. 

Export stream has an even better example. Export API allows 60 reqs/per hour and sparks with 3 reqs/per second.

So as long as we have fewer requests than the limit then we can make them all almost with no delay.

Reply:
Okay, can you leave a comment at the class level explaining this? This isn't very intuitive. Most developers would expect queries per second and not queries per hour.

Reply:
added